File: sliding-window/424-longest-repeating-char/424-ver-2.js
/**
 * @param {string} s
 * @param {number} k
 * @return {number}
 */
var characterReplacement = function (s, k) {
  let map = {};
  let maxCount = 0;
  let start = 0;
  let maxLen = 0;
  for (let end = 0; end < s.length; end++) {
    map[s[end]] = (map[s[end]] || 0) + 1;
    maxCount = maxCount > map[s[end]] ? maxCount : map[s[end]];
    // if meet the first char that is different from the previous
    // then end - start + 1 will be greater than maxCount;
    // if end - start + 1 - maxCount > k, then means there is no more
    // available flip;
    // then we need to shrink the window
    // and decrease the value of the removed char (which is the first letter at start)
    if (end - start + 1 - maxCount > k) {
      map[s[start]] -= 1;
      start++;
    }
    maxLen = maxLen > end - start + 1 ? maxLen : end - start + 1;
  }
};

File: sliding-window/424-longest-repeating-char/424-ver2.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        charSet = {}
        maxLen = 0
        left = 0
        largeCount = 0
        for right in range(len(s)):
            
            # create the dictionary to store the char and its count
            if s[right] not in charSet: charSet[s[right]] = 1
            else: charSet[s[right]] += 1

            # other way for the above:
            # charSet[s[right]] = 1 + charSet.get(s[right], 0)
            
            # store the largest count so far
            largeCount = max(largeCount, charSet[s[right]])
            
            # if the current length of the str - largest count > k
            # which means the difference str count is greater then
            # the times that we can flip the char
            # then we need to remove the left most char from the set
            if right + 1 - left - largeCount > k:
                charSet[s[left]] -= 1
                left += 1
            
            maxLen = max(right - left + 1, maxLen)
        return maxLen
                
                
            
            